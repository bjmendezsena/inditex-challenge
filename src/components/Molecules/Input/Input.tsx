import { useFormControlContext as useChakraFormControlContext } from "@chakra-ui/form-control";
import type { InputProps as ChakraInputProps } from "@chakra-ui/input";
import {
  Input as ChakraInput,
  InputGroup as ChakraInputGroup,
} from "@chakra-ui/input";
import type {
  SystemStyleObject as ChakraSystemStyleObject,
  ComponentWithAs,
} from "@chakra-ui/system";
import {
  forwardRef,
  useMultiStyleConfig as useChakraMultiStyleConfig,
} from "@chakra-ui/system";
import React from "react";
import _merge from "lodash/merge";
import { Text } from "../../Atoms";

import { Box as ChakraBox, useBreakpointValue } from "@chakra-ui/react";

import type { BaseMultiStyleConfiguration } from "../../../theme";

export const inputPartsList: Array<
  | "labelContainer"
  | "label"
  | "addonLeft"
  | "addonRight"
  | "input"
  | "container"
  | "errorIcon"
> = [
  "container",
  "addonLeft",
  "addonRight",
  "labelContainer",
  "label",
  "input",
  "errorIcon",
];

type Parts = typeof inputPartsList;

export type InputConfigProps = {
  size: string;
  variant: ChakraInputProps["variant"];
  hasLabel?: boolean;

  onClear?: React.MouseEventHandler<HTMLButtonElement>;
};
export interface InputThemeConfiguration
  extends BaseMultiStyleConfiguration<Parts> {}

export interface InputProps extends ChakraInputProps {
  size?: string;
  label?: string;
  onClear?: React.MouseEventHandler<HTMLButtonElement>;
}

export type InputThemeStyle = Record<Parts[number], ChakraSystemStyleObject> & {
  errorIcon: React.FunctionComponent;
  validIcon: React.FunctionComponent;
  clearIconButton: React.FunctionComponent;
};

export const defaultInputSize = { base: "lg", lg: "md" };
export const Input: ComponentWithAs<"input", InputProps> = forwardRef<
  InputProps,
  "input"
>(
  (
    {
      size,
      variant,
      id,
      isInvalid,
      isDisabled,
      label,
      placeholder = " ",
      onClear,
      sx,
      ...props
    }: InputProps,
    ref
  ) => {
    const formControl = useChakraFormControlContext() || {};
    const invalid = isInvalid ?? formControl.isInvalid;

    const defaultSize = useBreakpointValue(defaultInputSize);
    const inputSize = size || defaultSize;

    const styles = useChakraMultiStyleConfig("Input", {
      size: inputSize,
      variant,
      isDisabled,
      hasLabel: Boolean(label),
      onClear,
    }) as InputThemeStyle;

    const inputStyle = React.useMemo<ChakraSystemStyleObject>(
      () => _merge(styles.input, sx),
      [styles.input, sx]
    );

    return (
      <ChakraBox
        as='label'
        sx={styles.labelContainer}
        htmlFor={id}
        flex={1}
        position='relative'
        ref={ref}
      >
        {label && (
          <Text as='span' id='label' display='flex' sx={styles.label}>
            {label}
          </Text>
        )}
        <ChakraInputGroup overflow='hidden' sx={styles.container}>
          <ChakraInput
            size={inputSize}
            variant={variant}
            id={id}
            sx={inputStyle}
            isDisabled={isDisabled}
            isInvalid={invalid}
            placeholder={placeholder}
            {...props}
          />
        </ChakraInputGroup>
      </ChakraBox>
    );
  }
);
