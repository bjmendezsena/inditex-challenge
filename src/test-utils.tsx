import { ChakraProvider, extendTheme } from "@chakra-ui/react";
import { Router } from "react-router-dom";
import { createMemoryHistory, MemoryHistory } from "history";
import { QueryClient, QueryClientProvider } from "@tanstack/react-query";
import type { RenderResult } from "@testing-library/react";
import { render as testingLibraryRender } from "@testing-library/react";
import React from "react";
import { theme } from "./theme";

export const testQueryClient = new QueryClient({
  defaultOptions: {
    queries: {
      retry: false,
      refetchOnMount: true,
      cacheTime: 0,
    },
  },
  logger: {
    log: console.log,
    warn: console.warn,
    // âœ… no more errors on the console for tests
    error: process.env.NODE_ENV === "test" ? () => {} : console.error,
  },
});

export const queryWrapper = ({ children }: { children: React.ReactNode }) => (
  <QueryClientProvider client={testQueryClient}>{children}</QueryClientProvider>
);

const DefaultTheme = extendTheme(theme);

const memoryHistory = createMemoryHistory();

interface RenderOptions {
  history?: MemoryHistory;
}

export const render = (
  ui: JSX.Element,
  { history: customHistory }: RenderOptions = {}
): RenderResult => {
  const history = customHistory || memoryHistory;
  const wrapper = ({ children }: { children?: React.ReactNode }) => (
    <QueryClientProvider client={testQueryClient}>
      <ChakraProvider theme={DefaultTheme}>
        <Router location={history.location} navigator={history}>
          {children}
        </Router>
      </ChakraProvider>
    </QueryClientProvider>
  );

  return testingLibraryRender(ui, { wrapper });
};
