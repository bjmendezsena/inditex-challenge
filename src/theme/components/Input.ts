import { inputPartsList, type InputThemeConfiguration } from "../../components";

const baseStyle: InputThemeConfiguration["baseStyle"] = () => ({
  labelContainer: {
    width: "100%",
    position: "relative",
  },
  input: {
    paddingStart: 3,
    paddingEnd: 3,
    position: "relative",
    apply: "textStyles.ElementsRegularSM",
    color: "grey.800",
    textOverflow: "ellipsis",
    _placeholder: {
      apply: "textStyles.ElementsRegularSM",
      color: "grey.400",
    },
  },
  label: {
    apply: "textStyles.ElementsRegularSM",
    color: "grey.700",
    zIndex: 1,
    mb: 2,
  },
});

const sizesStyles = {
  sm: {
    input: {
      minH: 8,
      paddingY: "5px",
    },
  },
  md: {
    input: {
      minH: 10,
      paddingY: "9px",
    },
  },
  lg: {
    input: {
      minH: 12,
      paddingY: "13px",
    },
  },
};

const outline = () => ({
  input: {
    border: "1px solid",
    borderColor: "grey.300",
    borderRadius: "simple",
    _hover: {
      borderColor: "grey.500",
    },
    _focus: {
      borderColor: "primary",
      _hover: { borderColor: "primary" },
    },
    _active: {
      borderColor: "primary",
      _hover: { borderColor: "primary" },
    },
    _invalid: {
      borderColor: "error",
      _focus: {
        borderColor: "error",
        boxShadow: "none",
      },
      _active: {
        borderColor: "error",
        boxShadow: "none",
      },
    },
    _disabled: {
      borderColor: "grey.100",
      bg: "grey.50",
      color: "grey.300",
    },
  },
});

export const ConfigInput: InputThemeConfiguration = {
  parts: inputPartsList,
  baseStyle,
  sizes: {
    sm: sizesStyles.sm,
    md: sizesStyles.md,
    lg: sizesStyles.lg,
  },
  variants: {
    outline,
  },
  defaultProps: {
    variant: "outline",
  },
};
