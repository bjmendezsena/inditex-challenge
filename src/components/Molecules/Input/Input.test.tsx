import { render } from "../../../test-utils";
import type { InputProps } from "./Input";
import { Input } from "./Input";

describe(`<Input />`, () => {
  const id = "test-input";
  const label = "label";
  const placeholder = "placeholder";
  const defaultProps = {
    id,
    label,
    placeholder,
  };

  const isDisabled = { isDisabled: true };

  const factoryComponent = (props: InputProps = defaultProps) =>
    render(<Input {...props} />);

  test("renders correctly", () => {
    const view = factoryComponent();
    expect(view.asFragment()).toMatchSnapshot();
  });


  test("shows correctly when is disabled", () => {
    const view = factoryComponent({ ...defaultProps, ...isDisabled });
    expect(view.asFragment()).toMatchSnapshot("Using as disabled");
  });
});
