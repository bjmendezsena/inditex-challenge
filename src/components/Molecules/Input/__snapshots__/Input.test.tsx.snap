// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> renders correctly 1`] = `
<DocumentFragment>
  <label
    class="css-mimi7r"
    for="test-input"
  >
    <span
      class="css-b977hv"
      id="label"
    >
      label
    </span>
    <div
      class="chakra-input__group css-pwmwmr"
      data-group="true"
    >
      <input
        class="chakra-input css-1n414k7"
        id="test-input"
        placeholder="placeholder"
      />
    </div>
  </label>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;

exports[`<Input /> shows correctly when is disabled: Using as disabled 1`] = `
<DocumentFragment>
  <label
    class="css-mimi7r"
    for="test-input"
  >
    <span
      class="css-b977hv"
      id="label"
    >
      label
    </span>
    <div
      class="chakra-input__group css-pwmwmr"
      data-group="true"
    >
      <input
        class="chakra-input css-1n414k7"
        disabled=""
        id="test-input"
        placeholder="placeholder"
      />
    </div>
  </label>
  <span
    hidden=""
    id="__chakra_env"
  />
</DocumentFragment>
`;
